File: src/main/python/net/coderextreme/data/fors2.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "creator"
meta2.content = "John W Carlson"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "created"
meta3.content = "December 13 2015"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "modified"
meta4.content = "Mon, 09 Feb 2026 07:11:19 GMT"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "title"
meta5.content = "fors2.x3d"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "identifier"
meta6.content = "https://coderextreme.net/X3DJSONLD/src/main/data/fors2.x3d"

head1.children.append(meta6)

X3D0.head = head1
Scene7 = x3d.Scene()
ProtoDeclare8 = x3d.ProtoDeclare()
ProtoDeclare8.name = "node"
ProtoInterface9 = x3d.ProtoInterface()
field10 = x3d.field()
field10.accessType = "inputOutput"
field10.type = "SFVec3f"
field10.name = "position"

ProtoInterface9.field.append(field10)

ProtoDeclare8.ProtoInterface = ProtoInterface9
ProtoBody11 = x3d.ProtoBody()
Group12 = x3d.Group()
Transform13 = x3d.Transform(DEF="transform")
Shape14 = x3d.Shape()
Appearance15 = x3d.Appearance()
Material16 = x3d.Material()
Material16.diffuseColor = [1,0,0]

Appearance15.material = Material16

Shape14.appearance = Appearance15
Sphere17 = x3d.Sphere()

Shape14.geometry = Sphere17

Transform13.children.append(Shape14)
IS18 = x3d.IS()
connect19 = x3d.connect()
connect19.nodeField = "translation"
connect19.protoField = "position"

IS18.connect.append(connect19)

Transform13.IS = IS18

Group12.children.append(Transform13)
PositionInterpolator20 = x3d.PositionInterpolator(DEF="NodePosition")
PositionInterpolator20.key = [0,1]
PositionInterpolator20.keyValue = [(0, 0, 0),(0, 5, 0)]

Group12.children.append(PositionInterpolator20)
Script21 = x3d.Script(DEF="MoveBall")
field22 = x3d.field()
field22.accessType = "inputOutput"
field22.type = "SFVec3f"
field22.name = "translation"
field22.value = [50,50,0]

Script21.field.append(field22)
field23 = x3d.field()
field23.accessType = "inputOutput"
field23.type = "SFVec3f"
field23.name = "old"

Script21.field.append(field23)
field24 = x3d.field()
field24.accessType = "inputOnly"
field24.type = "SFTime"
field24.name = "set_cycle"

Script21.field.append(field24)
field25 = x3d.field()
field25.accessType = "outputOnly"
field25.type = "MFVec3f"
field25.name = "keyValue"

Script21.field.append(field25)

Script21.sourceCode = '''ecmascript:\n"+
"			function set_cycle(value) {\n"+
"				old = translation;\n"+
"				translation = new SFVec3f(Math.random()*100-50, Math.random()*100-50, Math.random()*100-50);\n"+
"				keyValue = new MFVec3f(old, translation);\n"+
"			}'''

Group12.children.append(Script21)
TimeSensor26 = x3d.TimeSensor(DEF="nodeClock")
TimeSensor26.cycleInterval = 3
TimeSensor26.loop = True

Group12.children.append(TimeSensor26)

ProtoBody11.children.append(Group12)
ROUTE27 = x3d.ROUTE()
ROUTE27.fromNode = "nodeClock"
ROUTE27.fromField = "cycleTime"
ROUTE27.toNode = "MoveBall"
ROUTE27.toField = "set_cycle"

ProtoBody11.children.append(ROUTE27)
ROUTE28 = x3d.ROUTE()
ROUTE28.fromNode = "nodeClock"
ROUTE28.fromField = "fraction_changed"
ROUTE28.toNode = "NodePosition"
ROUTE28.toField = "set_fraction"

ProtoBody11.children.append(ROUTE28)
ROUTE29 = x3d.ROUTE()
ROUTE29.fromNode = "MoveBall"
ROUTE29.fromField = "keyValue"
ROUTE29.toNode = "NodePosition"
ROUTE29.toField = "set_keyValue"

ProtoBody11.children.append(ROUTE29)
ROUTE30 = x3d.ROUTE()
ROUTE30.fromNode = "NodePosition"
ROUTE30.fromField = "value_changed"
ROUTE30.toNode = "transform"
ROUTE30.toField = "set_translation"

ProtoBody11.children.append(ROUTE30)

ProtoDeclare8.ProtoBody = ProtoBody11

Scene7.children.append(ProtoDeclare8)
ProtoDeclare31 = x3d.ProtoDeclare()
ProtoDeclare31.name = "cyl"
ProtoInterface32 = x3d.ProtoInterface()
field33 = x3d.field()
field33.accessType = "inputOutput"
field33.type = "SFVec3f"
field33.name = "positionA"

ProtoInterface32.field.append(field33)
field34 = x3d.field()
field34.accessType = "inputOutput"
field34.type = "SFVec3f"
field34.name = "positionB"

ProtoInterface32.field.append(field34)

ProtoDeclare31.ProtoInterface = ProtoInterface32
ProtoBody35 = x3d.ProtoBody()
Group36 = x3d.Group()
Shape37 = x3d.Shape()
Appearance38 = x3d.Appearance()
Material39 = x3d.Material()
Material39.diffuseColor = [0,1,0]

Appearance38.material = Material39

Shape37.appearance = Appearance38
Extrusion40 = x3d.Extrusion(DEF="extrusion")
Extrusion40.creaseAngle = 0.785
Extrusion40.crossSection = [(1, 0),(0.92, -0.38),(0.71, -0.71),(0.38, -0.92),(0, -1),(-0.38, -0.92),(-0.71, -0.71),(-0.92, -0.38),(-1, 0),(-0.92, 0.38),(-0.71, 0.71),(-0.38, 0.92),(0, 1),(0.38, 0.92),(0.71, 0.71),(0.92, 0.38),(1, 0)]
Extrusion40.spine = [(0, -50, 0),(0, 0, 0),(0, 50, 0)]

Shape37.geometry = Extrusion40

Group36.children.append(Shape37)
Script41 = x3d.Script(DEF="MoveCylinder")
field42 = x3d.field()
field42.accessType = "inputOutput"
field42.type = "MFVec3f"
field42.name = "spine"
field42.value = [(0, -50, 0),(0, 0, 0),(0, 50, 0)]

Script41.field.append(field42)
field43 = x3d.field()
field43.accessType = "inputOutput"
field43.type = "SFVec3f"
field43.name = "endA"

Script41.field.append(field43)
field44 = x3d.field()
field44.accessType = "inputOutput"
field44.type = "SFVec3f"
field44.name = "endB"

Script41.field.append(field44)

Script41.sourceCode = '''ecmascript:\n"+
"			function set_endA(value) {\n"+
"				spine = new MFVec3f(value, spine[1]);\n"+
"			}\n"+
"\n"+
"			function set_endB(value) {\n"+
"				spine = new MFVec3f(spine[0], value);\n"+
"			}'''
IS45 = x3d.IS()
connect46 = x3d.connect()
connect46.nodeField = "endA"
connect46.protoField = "positionA"

IS45.connect.append(connect46)
connect47 = x3d.connect()
connect47.nodeField = "endB"
connect47.protoField = "positionB"

IS45.connect.append(connect47)

Script41.IS = IS45

Group36.children.append(Script41)

ProtoBody35.children.append(Group36)
ROUTE48 = x3d.ROUTE()
ROUTE48.fromNode = "MoveCylinder"
ROUTE48.fromField = "spine_changed"
ROUTE48.toNode = "extrusion"
ROUTE48.toField = "set_spine"

ProtoBody35.children.append(ROUTE48)

ProtoDeclare31.ProtoBody = ProtoBody35

Scene7.children.append(ProtoDeclare31)
Transform49 = x3d.Transform()
Transform49.scale = [0.1,0.1,0.1]
ProtoInstance50 = x3d.ProtoInstance(DEF="nodeA")
ProtoInstance50.name = "node"
fieldValue51 = x3d.fieldValue()
fieldValue51.name = "position"
fieldValue51.value = "0 1.965 0"

ProtoInstance50.fieldValue.append(fieldValue51)

Transform49.children.append(ProtoInstance50)
ProtoInstance52 = x3d.ProtoInstance(DEF="nodeB")
ProtoInstance52.name = "node"
fieldValue53 = x3d.fieldValue()
fieldValue53.name = "position"
fieldValue53.value = "0 1.965 0"

ProtoInstance52.fieldValue.append(fieldValue53)

Transform49.children.append(ProtoInstance52)
ProtoInstance54 = x3d.ProtoInstance(DEF="linkA")
ProtoInstance54.name = "cyl"
fieldValue55 = x3d.fieldValue()
fieldValue55.name = "positionA"
fieldValue55.value = "0 1.965 0"

ProtoInstance54.fieldValue.append(fieldValue55)
fieldValue56 = x3d.fieldValue()
fieldValue56.name = "positionB"
fieldValue56.value = "0 1.965 0"

ProtoInstance54.fieldValue.append(fieldValue56)

Transform49.children.append(ProtoInstance54)

Scene7.children.append(Transform49)
ROUTE57 = x3d.ROUTE()
ROUTE57.fromNode = "nodeA"
ROUTE57.fromField = "position_changed"
ROUTE57.toNode = "linkA"
ROUTE57.toField = "set_positionA"

Scene7.children.append(ROUTE57)
ROUTE58 = x3d.ROUTE()
ROUTE58.fromNode = "nodeB"
ROUTE58.fromField = "position_changed"
ROUTE58.toNode = "linkA"
ROUTE58.toField = "set_positionB"

Scene7.children.append(ROUTE58)

X3D0.Scene = Scene7
f = open("../data/fors2.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/fors2.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/fors2.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/cplusplus/net/coderextreme/data/fors2.cpp
/* delete next 5 lines for Visual Studio */
#define FALSE false
#define TRUE true
#define WINAPI
#define AFX_EXT_CLASS
#define EXPORT32
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/CylinderExample/pch.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/CylinderExample/framework.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/XML_PARSER.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/X3DLib.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/Abstracts.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/Concretes.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/glMath.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/glut.h"
int main(int argc, char ** argv) {
X3D* X3D0 = new X3D();
X3D0->setProfile("Immersive");
X3D0->setVersion("3.3");
head* head1 = new head();
meta* meta2 = new meta();
meta2->setName("creator");
meta2->setContent("John W Carlson");
head1->addMeta(meta2);

meta* meta3 = new meta();
meta3->setName("created");
meta3->setContent("December 13 2015");
head1->addMeta(meta3);

meta* meta4 = new meta();
meta4->setName("modified");
meta4->setContent("April 18 2017");
head1->addMeta(meta4);

meta* meta5 = new meta();
meta5->setName("title");
meta5->setContent("fors2.x3d");
head1->addMeta(meta5);

meta* meta6 = new meta();
meta6->setName("identifier");
meta6->setContent("https://coderextreme.net/X3DJSONLD/src/main/data/fors2.x3d");
head1->addMeta(meta6);

meta* meta7 = new meta();
meta7->setName("description");
meta7->setContent("beginnings of a force directed graph in 3D");
head1->addMeta(meta7);

meta* meta8 = new meta();
meta8->setName("generator");
meta8->setContent("Vim, X3D-Edit, https://savage.nps.edu/X3D-Edit");
head1->addMeta(meta8);

X3D0->setHead(head1);

Scene* Scene9 = new Scene();
ProtoDeclare* ProtoDeclare10 = new ProtoDeclare();
ProtoDeclare10->setName("node");
ProtoInterface* ProtoInterface11 = new ProtoInterface();
field* field12 = new field();
field12->setName("position");
field12->setAccessType("inputOutput");
field12->setType("SFVec3f");
field12->setValue("0 0 0");
ProtoInterface11->addField(field12);

ProtoDeclare10->setProtoInterface(ProtoInterface11);

ProtoBody* ProtoBody13 = new ProtoBody();
Transform* Transform14 = new Transform();
Transform14->setDEF("transform");
IS* IS15 = new IS();
Connect* connect16 = new Connect();
connect16->setNodeField("translation");
connect16->setProtoField("position");
IS15->addConnect(connect16);

Transform14->setIS(IS15);

Shape* Shape17 = new Shape();
//comment before Sphere
//comment after Sphere
//comment after Appearance
Sphere* Sphere18 = new Sphere();
Shape17->setGeometry(Sphere18);

Appearance* Appearance19 = new Appearance();
//comment before Material
//comment after Material
Material* Material20 = new Material();
Material20->setDiffuseColor(new float[3]{1,0,0});
Appearance19->setMaterial(Material20);

Shape17->setAppearance(Appearance19);

Transform14->addChild(Shape17);

ProtoBody13->addChild(Transform14);

PositionInterpolator* PositionInterpolator21 = new PositionInterpolator();
PositionInterpolator21->setDEF("NodePosition");
PositionInterpolator21->setKey(new float[2]{0,1}, 2);
PositionInterpolator21->setKeyValue(new float[6]{0,0,0,0,5,0});
ProtoBody13->addChild(PositionInterpolator21);

Script* Script22 = new Script();
Script22->setDEF("MoveBall");
field* field23 = new field();
field23->setName("translation");
field23->setAccessType("inputOutput");
field23->setType("SFVec3f");
field23->setValue("50 50 0");
Script22->addField(field23);

field* field24 = new field();
field24->setName("old");
field24->setAccessType("inputOutput");
field24->setType("SFVec3f");
field24->setValue("0 0 0");
Script22->addField(field24);

field* field25 = new field();
field25->setName("set_cycle");
field25->setAccessType("inputOnly");
field25->setType("SFTime");
Script22->addField(field25);

field* field26 = new field();
field26->setName("keyValue");
field26->setAccessType("outputOnly");
field26->setType("MFVec3f");
Script22->addField(field26);


Script22.setSourceCode(`ecmascript:\n"+
"					function set_cycle(value) {\n"+
"                                                old = translation;\n"+
"						translation = new SFVec3f(Math.random()*100-50, Math.random()*100-50, Math.random()*100-50);\n"+
"                                                keyValue = new MFVec3f([old, translation]);\n"+
"						// Browser.println(translation);\n"+
"					}`)
ProtoBody13->addChild(Script22);

TimeSensor* TimeSensor27 = new TimeSensor();
TimeSensor27->setDEF("nodeClock");
TimeSensor27->setCycleInterval(3);
TimeSensor27->setLoop(True);
ProtoBody13->addChild(TimeSensor27);

ROUTE* ROUTE28 = new ROUTE();
ROUTE28->setFromNode("nodeClock");
ROUTE28->setFromField("cycleTime");
ROUTE28->setToNode("MoveBall");
ROUTE28->setToField("set_cycle");
ProtoBody13->addChild(ROUTE28);

ROUTE* ROUTE29 = new ROUTE();
ROUTE29->setFromNode("nodeClock");
ROUTE29->setFromField("fraction_changed");
ROUTE29->setToNode("NodePosition");
ROUTE29->setToField("set_fraction");
ProtoBody13->addChild(ROUTE29);

ROUTE* ROUTE30 = new ROUTE();
ROUTE30->setFromNode("MoveBall");
ROUTE30->setFromField("keyValue");
ROUTE30->setToNode("NodePosition");
ROUTE30->setToField("keyValue");
ProtoBody13->addChild(ROUTE30);

ROUTE* ROUTE31 = new ROUTE();
ROUTE31->setFromNode("NodePosition");
ROUTE31->setFromField("value_changed");
ROUTE31->setToNode("transform");
ROUTE31->setToField("set_translation");
ProtoBody13->addChild(ROUTE31);

ProtoDeclare10->setProtoBody(ProtoBody13);

Scene9->addChild(ProtoDeclare10);

ProtoDeclare* ProtoDeclare32 = new ProtoDeclare();
ProtoDeclare32->setName("cyl");
ProtoInterface* ProtoInterface33 = new ProtoInterface();
field* field34 = new field();
field34->setName("positionA");
field34->setAccessType("inputOnly");
field34->setType("SFVec3f");
ProtoInterface33->addField(field34);

field* field35 = new field();
field35->setName("positionB");
field35->setAccessType("inputOnly");
field35->setType("SFVec3f");
ProtoInterface33->addField(field35);

ProtoDeclare32->setProtoInterface(ProtoInterface33);

ProtoBody* ProtoBody36 = new ProtoBody();
Shape* Shape37 = new Shape();
Extrusion* Extrusion38 = new Extrusion();
Extrusion38->setDEF("extrusion");
Extrusion38->setCreaseAngle(0.785);
Extrusion38->setCrossSection(new float[34]{1,0,0.92,-0.38,0.71,-0.71,0.38,-0.92,0,-1,-0.38,-0.92,-0.71,-0.71,-0.92,-0.38,-1,0,-0.92,0.38,-0.71,0.71,-0.38,0.92,0,1,0.38,0.92,0.71,0.71,0.92,0.38,1,0});
Extrusion38->setSpine(new float[9]{0,-50,0,0,0,0,0,50,0});
Shape37->setGeometry(Extrusion38);

Appearance* Appearance39 = new Appearance();
Material* Material40 = new Material();
Material40->setDiffuseColor(new float[3]{0,1,0});
Appearance39->setMaterial(Material40);

Shape37->setAppearance(Appearance39);

ProtoBody36->addChild(Shape37);

Script* Script41 = new Script();
Script41->setDEF("MoveCylinder");
field* field42 = new field();
field42->setName("spine");
field42->setAccessType("inputOutput");
field42->setType("MFVec3f");
field42->setValue("0 -50 0 0 0 0 0 50 0");
Script41->addField(field42);

field* field43 = new field();
field43->setName("set_endA");
field43->setAccessType("inputOnly");
field43->setType("SFVec3f");
Script41->addField(field43);

field* field44 = new field();
field44->setName("set_endB");
field44->setAccessType("inputOnly");
field44->setType("SFVec3f");
Script41->addField(field44);

IS* IS45 = new IS();
Connect* connect46 = new Connect();
connect46->setNodeField("set_endA");
connect46->setProtoField("positionA");
IS45->addConnect(connect46);

Connect* connect47 = new Connect();
connect47->setNodeField("set_endB");
connect47->setProtoField("positionB");
IS45->addConnect(connect47);

Script41->setIS(IS45);


Script41.setSourceCode(`ecmascript:\n"+
"\n"+
"                function set_endA(value) {\n"+
"		    if (typeof spine === \"undefined\") {\n"+
"		        spine = new MFVec3f([value, value]);\n"+
"		    } else {\n"+
"		        spine = new MFVec3f([value, spine[1]]);\n"+
"		    }\n"+
"                }\n"+
"\n"+
"                function set_endB(value) {\n"+
"		    if (typeof spine === \"undefined\") {\n"+
"		        spine = new MFVec3f([value, value]);\n"+
"		    } else {\n"+
"		        spine = new MFVec3f([spine[0], value]);\n"+
"		    }\n"+
"                }\n"+
"\n"+
"                function set_spine(value) {\n"+
"		    Browser.print('\\n'+'\"');\n"+
"                    spine = value;\n"+
"                }`)
ProtoBody36->addChild(Script41);

ROUTE* ROUTE48 = new ROUTE();
ROUTE48->setFromNode("MoveCylinder");
ROUTE48->setFromField("spine");
ROUTE48->setToNode("extrusion");
ROUTE48->setToField("set_spine");
ProtoBody36->addChild(ROUTE48);

ProtoDeclare32->setProtoBody(ProtoBody36);

Scene9->addChild(ProtoDeclare32);

Transform* Transform49 = new Transform();
Transform49->setScale(new float[3]{0.1,0.1,0.1});
ProtoInstance* ProtoInstance50 = new ProtoInstance();
ProtoInstance50->setName("node");
ProtoInstance50->setDEF("nodeA");
fieldValue* fieldValue51 = new fieldValue();
fieldValue51->setName("position");
fieldValue51->setValue("-50 -50 -50");
ProtoInstance50->addFieldValue(fieldValue51);

Transform49->addChild(ProtoInstance50);

ProtoInstance* ProtoInstance52 = new ProtoInstance();
ProtoInstance52->setName("node");
ProtoInstance52->setDEF("nodeB");
fieldValue* fieldValue53 = new fieldValue();
fieldValue53->setName("position");
fieldValue53->setValue("50 50 50");
ProtoInstance52->addFieldValue(fieldValue53);

Transform49->addChild(ProtoInstance52);

ProtoInstance* ProtoInstance54 = new ProtoInstance();
ProtoInstance54->setName("cyl");
ProtoInstance54->setDEF("linkA");
fieldValue* fieldValue55 = new fieldValue();
fieldValue55->setName("positionA");
fieldValue55->setValue("0 0 0");
ProtoInstance54->addFieldValue(fieldValue55);

fieldValue* fieldValue56 = new fieldValue();
fieldValue56->setName("positionB");
fieldValue56->setValue("50 50 50");
ProtoInstance54->addFieldValue(fieldValue56);

Transform49->addChild(ProtoInstance54);

Scene9->addChild(Transform49);

ROUTE* ROUTE57 = new ROUTE();
ROUTE57->setFromNode("nodeA");
ROUTE57->setFromField("position");
ROUTE57->setToNode("linkA");
ROUTE57->setToField("positionA");
Scene9->addChild(ROUTE57);

ROUTE* ROUTE58 = new ROUTE();
ROUTE58->setFromNode("nodeB");
ROUTE58->setFromField("position");
ROUTE58->setToNode("linkA");
ROUTE58->setToField("positionB");
Scene9->addChild(ROUTE58);

X3D0->setScene(Scene9);

X3D0->toXMLString();
}
